msvc/scripts/genconsts.cs

```diff
@@ -0,0 +1,98 @@
+using System;
+using System.IO;
+using System.Reflection;
+using System.Diagnostics;
+using System.Text;
+using System.Text.RegularExpressions;
+
+public static class Program {
+    public static int Main (string[] args) {
+        var myAssembly = Assembly.GetExecutingAssembly ();
+        var codeBase = new Uri (myAssembly.CodeBase);
+        var executablePath = Path.GetFullPath (codeBase.LocalPath);
+        var executableDirectory = Path.GetDirectoryName (executablePath);
+
+        var winsetupDirectory = Path.Combine (executableDirectory, "..");
+        var winsetupPath = Path.Combine (winsetupDirectory, "winsetup.bat");
+
+        var psi = new ProcessStartInfo (winsetupPath) {
+            WorkingDirectory = winsetupDirectory,
+            UseShellExecute = false,
+            ErrorDialog = false,
+            RedirectStandardOutput = true
+        };
+
+        string monoVersion, monoCorlibVersion;
+
+        Process winsetupProcess;
+
+        try {
+            winsetupProcess = Process.Start (psi);
+        } catch (Exception exc) {
+            Console.Error.WriteLine ("Failed starting winsetup.bat");
+            Console.Error.WriteLine (exc);
+            return 1;
+        }
+
+        using (winsetupProcess) {
+            var outputBuffer = new StringBuilder ();
+
+            winsetupProcess.OutputDataReceived += (s, e) => {
+                outputBuffer.AppendLine (e.Data);
+            };
+            winsetupProcess.BeginOutputReadLine ();
+            winsetupProcess.WaitForExit ();
+
+            var output = outputBuffer.ToString ().Trim ();
+
+            if (winsetupProcess.ExitCode != 0) {
+                Console.Error.WriteLine ("Failed running winsetup.bat");
+                Console.Write (output);
+                return winsetupProcess.ExitCode;
+            } else {
+                var m = Regex.Match (output, "MONO_VERSION=([0-9.]+)");
+                if (!m.Success)
+                    return 1;
+                monoVersion = m.Groups[1].Value;
+
+                // HACK: winsetup.bat produces N.N.N instead of N.N.N.N like configure.ac,
```

Comment:
Maybe add another line or two to winsetup.bat to remove the hack? i.e. to output another line with the data you need?

Reply:
Yeah, if that makes sense to you. I'm not sure why the batch file isn't generated, so I thought you might know.

Reply:
This is why btw, it is to mimic:
```
config.h:#define PACKAGE_VERSION "5.17.0"
config.h:#define VERSION "5.17.0"
```

Reply:
```
$ /s/mono/mono/mini/mono --version
Mono JIT compiler version 5.17.0 (cominterop.3/4660b820c27 Mon Jul 16 15:22:46 PDT 2018)

jaymac1:mono jay$ which mono
/Library/Frameworks/Mono.framework/Versions/Current/Commands/mono
jaymac1:mono jay$ mono --version
Mono JIT compiler version 5.10.1.47 (2017-12/8eb8f7d5e74 Fri Apr 13 20:18:12 EDT 2018)
```